src/preprocessing/shapes/polygon_shape.jl

```diff
@@ -0,0 +1,101 @@
+# This is the data format returned by `load(file)` when used with `.asc` files
+struct Polygon{NDIMS, ELTYPE} <: Shapes{NDIMS}
+    vertices       :: Vector{SVector{NDIMS, ELTYPE}}
+    edge_vertices  :: Vector{NTuple{2, SVector{NDIMS, ELTYPE}}}
+    vertex_normals :: Vector{Vector{SVector{NDIMS, ELTYPE}}}
+    edge_normals   :: Vector{SVector{NDIMS, ELTYPE}}
+    min_corner     :: SVector{NDIMS, ELTYPE}
+    max_corner     :: SVector{NDIMS, ELTYPE}
+
+    function Polygon(vertices)
+        NDIMS = size(vertices, 1)
+
+        # Close the polygon if it's open
+        if !isapprox(vertices[:, end], vertices[:, 1])
+            vertices = hcat(vertices, vertices[:, 1])
+        end
+
+        min_corner = SVector{NDIMS}(minimum(vertices, dims=2))
+        max_corner = SVector{NDIMS}(maximum(vertices, dims=2))
+
+        n_vertices = size(vertices, 2)
+        ELTYPE = eltype(vertices)
+
+        # Sum over all the edges and determine if the vertices are in clockwise order
+        # to make sure that all normals pointing outwards.
+        # To do so, we compute the area of the polygon, which can be either positive or
+        # negative. It depends on the order of the vertices.
+        # https://en.wikipedia.org/wiki/Polygon
+        polygon_area = 0.0
+        for i in 1:(n_vertices - 1)
+            v1 = extract_svector(vertices, Val(NDIMS), i)
+            v2 = extract_svector(vertices, Val(NDIMS), i + 1)
+            polygon_area += (v1[1] * v2[2] - v2[1] * v1[2])
+        end
+
+        if polygon_area == 0.0
+            throw(ArgumentError("polygon is not correctly defined"))
+        elseif polygon_area > 0.0
+            # Curve is counter-clockwise
+            reverse!(vertices, dims=2)
+        end
+
+        edge_vertices = Vector{NTuple{2, SVector{NDIMS, ELTYPE}}}()
+        edge_normals = Vector{SVector{NDIMS, ELTYPE}}()
+
+        for i in 1:(n_vertices - 1)
+            v1 = extract_svector(vertices, Val(NDIMS), i)
+            v2 = extract_svector(vertices, Val(NDIMS), i + 1)
+            if isapprox(v1, v2)
+                continue
+            end
+
+            edge = v2 - v1
+
+            edge_normal = SVector{NDIMS}(normalize([-edge[2], edge[1]]))
+
+            push!(edge_vertices, (v1, v2))
+            push!(edge_normals, edge_normal)
+        end
+
+        vertex_normals = Vector{Vector{SVector{NDIMS, ELTYPE}}}()
+
+        # Calculate vertex pseudo-normals
+        # A edge is defined with two vertices of which we calculate the pseudo-normals
+        # by averaging the normals of the neighbor edges.
```

Comment:
Yes, it will be useful for the signed distances. 
We calculate the distances for a single edge and don't look for adjacent edges. That means, we do not iterate over all edges.

Reply:
Then please add this in a comment.